doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,706 @@
+within NumPy they require increased complexity.
+In general the concerns of different datatypes are not well separated.
+This burden is exacerbated by the exposure of internal C structures,
+limiting for example the addition of new sorting methods.
+
+Currently there are many factors which limit the creation of new user defined
+datatypes:
+
+* Creating casting rules for parametric user types is either impossible
+  or so complex that it has never been attempted.
+* Type promotion, e.g. the operation deciding that adding float and integer
+  values should return a float value, is very valuable for numeric datatypes
+  but is limited in scope for user defined and especially parametric datatypes.
+* Much of the logic is written in single functions
+  instead of being split as methods on the datatype itself.
+* In the current design datatypes cannot have methods that do not generalize
+  to other datatypes. For example a unit datatype cannot have a ``.to_si()`` method to
+  easily find the datatype which would represent the same values in SI units.
+
+The large need to solve these issues has driven the scientific community
+to create work-arounds in multiple projects implementing physical units as an
+array-like class instead of a datatype, which would generalize better across
+multiple array-likes (Dask, pandas, etc.).
+Already, Pandas has made a push into the same direction with its
+extension arrays [pandas_extension_arrays]_ and undoubtedly
+the community would be best served if such new features could be common
+between NumPy, Pandas, and other projects.
+
+Scope
+^^^^^
+
+This NEP proposes to move ahead with the necessary creation of new dtype
+subclasses (see Phase I Implementation section),
+and start working on implementing current functionality within the new context
+Within the context of this NEP all development will be fully private API or
+use preliminary underscored names which must be changed in the future.
+Most of the internal and public API choices are part of a second Phase
+and will be discussed in more detail in the following NEPs 42 and 43.
+The initial implementation of this NEP will have little or no effect on users,
+but provides the necessary ground work for incrementally addressing the
+full rework.
+
+The implementation of this NEP and leading up to a large rework of how
+datatypes are defined in NumPy is expected to create small incompatibilities
+(see backward compatibility section).
+However, a transition requiring large code adaption is not anticipated and not
+within scope.
+
+Specifically, this NEP makes the following design choices which are discussed
+in more details in the detailed description section:
+
+1. Each datatype will be a class instance of a DType class, with most of the
+   datatype-specific logic being implemented
+   as special methods on the class. In the C-API, these correspond to specific
+   slots. In short ``type(np.dtype("f8"))`` will be a subclass of ``np.dtype``.
```

Comment:
```suggestion
   slots. In short, for ``f = np.dtype("f8")``, ``isinstance(f, np.dtype)`` will remain true, but ``type(f)`` will be a subclass of ``np.dtype`` rather than just ``np.dtype`` itself.
```